### PR TITLE
Convert from Webpack to ESBuild

### PR DESCRIPTION
**Pull request by Ferotiq**
### Description

Yet another bundler change.

The compile times are a wet dream, though.

Closes #2577

### CODE REVIEW

**Comment by Miodec** [comment]
What about circular dependencies check?

**Comment by Ferotiq** [comment]
we're still trying to figure that out.

File: frontend/src/scripts/ape/endpoints/users.ts [hunk 1]
```diff
@@ -33,9 +33,9 @@ export default function getUsersEndpoints(
     return await apeClient.patch(`${BASE_PATH}/name`, { payload: { name } });
   }
 
-  async function updateLeaderboardMemory(
-    mode: string,
-    mode2: MonkeyTypes.Mode2<any>,
+  async function updateLeaderboardMemory<M extends MonkeyTypes.Mode>(
```
Comment (Bruception) [inline frontend/src/scripts/ape/endpoints/users.ts:36]: Does this need to be updated? https://github.com/Miodec/monkeytype/blob/master/frontend/src/scripts/db.ts#L612

Comment (Ferotiq) [reply frontend/src/scripts/ape/endpoints/users.ts:36]: was getting compile errors

Comment (Ferotiq) [reply frontend/src/scripts/ape/endpoints/users.ts:36]: or something I'm not quite sure

File: backend/package.json [hunk 2]
```diff
@@ -45,7 +45,7 @@
     "@types/lodash": "4.14.178",
     "@types/mongodb": "4.0.7",
     "@types/node": "17.0.18",
-    "@types/swagger-stats": "0.95.4",
+    "@types/swagger-stats": "^0.95.4",
```
Comment (Bruception) [inline backend/package.json:48]: This needs to be reverted.

File: backend/package-lock.json [hunk 3]
```diff
@@ -38,7 +38,7 @@
         "@types/lodash": "4.14.178",
         "@types/mongodb": "4.0.7",
         "@types/node": "17.0.18",
-        "@types/swagger-stats": "0.95.4",
+        "@types/swagger-stats": "^0.95.4",
```
Comment (Bruception) [inline backend/package-lock.json:41]: Hmm somehow this is still messed up.

Comment (Ferotiq) [reply backend/package-lock.json:41]: bruh

**Comment by Ferotiq** [comment]
I just pinged typescript-eslint LMAO

File: frontend/src/scripts/popups/contact-popup.ts [hunk 4]
```diff
@@ -14,3 +14,5 @@ $(document.body).on("click", "#contactPopupWrapper", (e) => {
       });
   }
 });
+
+export {};
```
Comment (Bruception) [inline frontend/src/scripts/popups/contact-popup.ts:18]: Why this export?

Comment (Ferotiq) [reply frontend/src/scripts/popups/contact-popup.ts]: nevermind, I guess it doesn't do it anymore, reverted.

File: frontend/src/scripts/modules.d.ts [hunk 5]
```diff
@@ -1,4 +1,5 @@
-declare let firebase: any; // typeof import("firebase").default;
-declare let moment: typeof import("moment");
+import "grecaptcha";
+
+// // declare let firebase: any; // typeof import("firebase").default;
```
Comment (Bruception) [inline frontend/src/scripts/modules.d.ts:3]: Can we remove these comments?

Comment (Ferotiq) [reply frontend/src/scripts/modules.d.ts:3]: yeah

File: frontend/tsconfig.json [hunk 6]
```diff
@@ -17,7 +17,6 @@
     "allowUmdGlobalAccess": true /* Allow accessing UMD globals from modules. */,
     "resolveJsonModule": true /* Enable importing .json files */,
     "allowJs": true /* Allow JavaScript files to be a part of your program. Use the `checkJS` option to get errors from these files. */,
-    "outDir": "./build" /* Specify an output folder for all emitted files. */,
```
Comment (Bruception) [inline frontend/tsconfig.json:20]: I think we should keep this line, if we just want to do typescript compilation.
Does this conflict with the build process?

**Comment by Ferotiq** [comment]
Converting back to draft because we need to minify and do code-splitting, the minification relies on eval being removed from the commandline.

**Comment by Ferotiq** [comment]
### Final Compile Times
Gulp `build` task (webpack): 29s
Gulp `build` task (esbuild): 16s `45% less time`
Gulp `build-production` task (webpack): 47s
Gulp `build-production` task (esbuild): 33s `30% less time`
Github PR Check (webpack): 1m 22s
Github PR Check (esbuild): 54s `34% less time`

Our precious seconds will be saved.

File: frontend/package.json [hunk 7]
```diff
@@ -16,52 +22,52 @@
...
-    "webpack": "^5.68.0",
-    "webpack-cli": "4.9.2"
+    "del": "6.0.0",
+    "esbuild": "0.14.23",
+    "gulp": "4.0.2",
+    "gulp-concat": "2.6.1",
+    "gulp-eslint-new": "1.3.0",
+    "gulp-sass": "5.0.0",
+    "gulp-typescript": "6.0.0-alpha.1",
+    "madge": "^5.0.1",
```
Comment (Bruception) [inline frontend/package.json:43]: 🥕

File: frontend/package.json [hunk 8]
```diff
@@ -16,52 +22,52 @@
...
+    "axios": "0.21.2",
+    "chart.js": "2.9.4",
+    "chartjs-plugin-annotation": "0.5.7",
+    "chartjs-plugin-trendline": "0.2.2",
+    "crypto-browserify": "3.12.0",
+    "dom-to-image": "2.6.0",
+    "easing": "1.2.1",
+    "firebase": "8.4.2",
+    "grecaptcha": "1.0.3",
+    "gulp-babel": "^8.0.0",
```
Comment (Bruception) [inline frontend/package.json:60]: 🥕

File: frontend/src/scripts/misc.ts [hunk 9]
```diff
@@ -1,8 +1,10 @@
+import firebase from "firebase";
+import moment from "moment";
 import * as Loader from "./elements/loader";
 
 export function getuid(): void {
   console.error("Only share this uid with Miodec and nobody else!");
```
Comment (Bruception) [inline frontend/src/scripts/misc.ts:6]: I wonder why this comment. UID should be safe to expose

File: frontend/tsconfig.json [hunk 10]
```diff
@@ -1,23 +1,24 @@
...
   "compilerOptions": {
-    "target": "es6" /* Set the JavaScript language version for emitted JavaScript and include compatible library declarations. */,
+    "target": "ES6" /* Set the JavaScript language version for emitted JavaScript and include compatible library declarations. */,
     "lib": [
       "ESNext",
       "DOM",
       "DOM.Iterable"
     ] /* Specify a set of bundled library declaration files that describe the target runtime environment. */,
-    "module": "ES6" /* Specify what module code is generated. */,
+    "module": "AMD" /* Specify what module code is generated. */,
```
Comment (Bruception) [inline frontend/tsconfig.json:9]: Reasoning for this change?

File: package.json [hunk 11]
```diff
@@ -25,8 +25,8 @@
     "npm": "8.1.2"
   },
   "devDependencies": {
-    "@typescript-eslint/eslint-plugin": "^5.11.0",
-    "@typescript-eslint/parser": "^5.11.0",
+    "@typescript-eslint/eslint-plugin": "^5.13.0",
```
Comment (Bruception) [inline package.json:28]: 🥕

**Comment by Bruception** [comment]
LGTM IIW, just some comments

File: frontend/tsconfig.json (hunk 10 above)
Comment (Ferotiq) [reply frontend/tsconfig.json]: I had some issue with it before, but I tried it now and it works so I think it's fine to remove.

File: frontend/src/scripts/misc.ts (hunk 9 above)
Comment (Ferotiq) [reply frontend/src/scripts/misc.ts:6]: 🤷 ask mio

**Comment by Ferotiq** [comment]
I’ll draft this again

**Pull request closed without merging**